21861 dashboard filter in db and cards sla per tech group tech

btry opened the pull request:
- migrates dashboard filters to database
- add a private mode to dashboards. A private is invisible to other users except if it is explicitly shared.
- adds 2 dashboard cards to see 
  - count of tickets by SLA status and technician
  - count of tickets by SLA status and technician group

- adds 2 new graph types
  -  multiple horizontal bars
  - horizontal stacked bars

![image](https://user-images.githubusercontent.com/14139801/122534711-1504c900-d023-11eb-8da6-ce410ba4365d.png)


Private mode 
![image](https://user-images.githubusercontent.com/14139801/123618679-b7cdfc00-d808-11eb-8ed0-d1d0e5976959.png)

btry commented on inc/dashboard/filter.class.php:
```diff
@@ -164,7 +167,7 @@ static function user_tech(string $value = ""): string {
    }
 
-   static function dropdown(
+   static function itemDropdown(
```
Method renamed because of a conflict with the implementation of the ancestor. Before  this PR, the class does not have any ancestor.

I dislike this change, but it is necessary. Maybe I can improve it by finding a better name.

btry commented on inc/dashboard/provider.class.php:
```diff
@@ -413,6 +416,317 @@ static function nbTicketsGeneric(
+         ]
+      ];
+
+      unset($params['apply_filters']['group_tech']);
```
this card is not compatible with group_tech filter. To prevent false results, it is necessary to not apply the filter.

I did not find a proper way to do this. Comments welcome

btry commented on inc/dashboard/provider.class.php:
```diff
@@ -413,6 +416,317 @@ static function nbTicketsGeneric(
+         ]
+      ];
+
+      unset($params['apply_filters']['user_tech']);
```
this card is not compatible with user_tech filter. To prevent false results, it is necessary to not apply the filter.

I did not find a proper way to do this. Comments welcome

btry commented on inc/dashboard/widget.class.php:
```diff
@@ -878,20 +878,21 @@ private static function getBarsGraph(
+         'limit'             => 99999,
+         'filters'           => [],
+         'rand'              => mt_rand(),
+         'truncate_boundary' => 'beginning',
```
new option : the 2 new dashboard cards rely on ordering to show the most important data. However, the code truncates the data on the wrong bound (because it was designed to work on data ordered by date / time).

orthagh commented:
## functionnal review

**Regarding the private dashboard feature:**

- private field should be a input[type=checkbox] available on dashboard creation:
![image](https://user-images.githubusercontent.com/418844/122392131-e67be500-cf73-11eb-971a-d269a8665790.png)

- when the dashboard already exists, i think the control to toggle the privacy should be in the share modal:

![image](https://user-images.githubusercontent.com/418844/122392373-1b883780-cf74-11eb-8b62-94ab4c24b467.png)

when the "is private" is check, the share with user/group/entities should be disabled (grayed out).
I think adding a dedicated control in the dashboard toolbar is not necessary and pollute the UX

- In both place, the field should have an helper (icon displaying a tooltip) to explains inline what set a dashboard as private means.


**Regarding the filter in db:** nothing to say on functional side

**Regarding the new cards:**

I'll talk about the two card, as they display very similar data.
i think the current two widget options are not very convenient as the presented data are not time based.
I think, in this particular case, a vertical alternative for stackedBar and Multiplebar widget could be better

example with sla/status/group card (blue part is an obsfucation of data):
![image](https://user-images.githubusercontent.com/418844/122393757-81c18a00-cf75-11eb-8af0-1113974eacde.png)

The addition of these two new widgets should be easy i think, it should consist to:

- add a new dedicated function:
```php
   public static function horizontalStackedBars(array $params = []): string {
      return self::StackedBars(array_merge($params, [
         'horizontal' => true,
      ]));
   }
```

- declare the functions in `Widget::getAllTypes`
- use it the card declaration (in `Grid::getAllDasboardCards`)
- add picture for new widget (maybe the hardest part 😁 )

orthagh commented on inc/dashboard/dashboard.class.php:
```diff
@@ -487,4 +519,22 @@ static function importFromJson($import = null) {
+
+      return $this->update([
+         'id'       => $this->fields['id'],
+         'key'     => $this->fields['key'],
```
```suggestion
         'key'      => $this->fields['key'],
```

orthagh commented on inc/dashboard/filter.class.php:
```diff
@@ -255,4 +258,47 @@ static function field(
    }
+
+   /**
+    * Return rights for the provided dashboard
```
```suggestion
    * Return filters for the provided dashboard
```

orthagh commented on inc/dashboard/provider.class.php:
```diff
@@ -413,6 +414,301 @@ static function nbTicketsGeneric(
+
+      $default_params = [
+         'label'         => "",
+         'icon'          => 'far fa-clock',
```
```suggestion
         'icon'          => 'fas fa-stopwatch',
```

orthagh commented on inc/dashboard/provider.class.php:
```diff
@@ -413,6 +414,301 @@ static function nbTicketsGeneric(
+
+      $default_params = [
+         'label'         => "",
+         'icon'          => 'far fa-clock',
```
```suggestion
         'icon'          => 'fas fa-stopwatch',
```

btry commented:
@orthagh : I changed the cards to use only horizontal bars (multiple or stacked). In this mode, boundary option I introduced truncates the data on the wrong boundary, then I reverted it.

Code cleanup, method naming are fixed. I'll work soon on the private feature to apply UI changes. If you have opportunity, please see if the 2 new horizontal display modes are working for you.

btry commented on inc/dashboard/dashboard.class.php:
```diff
@@ -426,11 +456,6 @@ static function convertRights(array $raw_rights = []): array {
-      if (self::canView()) {
-         return true;
-      }
-
```
To be checked carefully : rights checks are spread on several places.

Here I removed the canView() call because CheckRights is called in CanViewCurrent(). This method already calls canView().

In getAll() I replaced checkRights() by a call to canViewCurrent() to properly check the private status. This changes makes taht the checks includes the view all dashboards right, which was not checked before. I expect validation from reviewers.

Finally, with all the right check changes, the method checkRights may be merged with canViewCurrent() or renamed into checkShareRights().

btry replied:
ping @orthagh

orthagh replied:
should be ok

orthagh reviewed (approved):
ok for me.

Need at least : 

- reword of private in "Personal"
- some tests on new providers

orthagh reviewed (approved):
ok for merge

orthagh commented on tests/units/Glpi/Dashboard/Provider.php:
```diff
@@ -0,0 +1,293 @@
+        $this->integer($output['data']['series'][2]['data'][0])->isEqualTo(0);
+        $this->integer($output['data']['series'][3]['data'][0])->isEqualTo(0);
+    }
+}
```
```suggestion
}

```

cedric-anne commented on inc/dashboard/filter.class.php:
```diff
@@ -255,4 +258,47 @@ static function field(
+         ]
+      ]);
+
+      $settings = $dr_iterator->next()['filter'];
```
Unless I am wrong, filter can be missing in DB. For instance, migration does not create a filter for every existing dashboard.
```suggestion
      $settings = $dr_iterator->count() === 1 ? $dr_iterator->next()['filter'] : null;
```

cedric-anne commented on js/dashboard.js:
```diff
@@ -396,9 +400,10 @@ var Dashboard = {
          delete filters[filter_id];
-         Dashboard.setFiltersInStorage(filters);
+         Dashboard.setFiltersInDB(filters);
+         //Dashboard.setFiltersInStorage(filters);
```
```suggestion
```

cedric-anne commented on js/dashboard.js:
```diff
@@ -659,9 +664,10 @@ var Dashboard = {
       filters[filter_id] = value;
-      Dashboard.setFiltersInStorage(filters);
+      Dashboard.setFiltersInDB(filters);
+      //Dashboard.setFiltersInStorage(filters);
```
```suggestion
```

cedric-anne commented on js/dashboard.js:
```diff
@@ -1043,64 +1049,63 @@ var Dashboard = {
 
-            Dashboard.setFiltersInStorage(new_filters);
+            Dashboard.setFiltersInDB(new_filters);
+            //Dashboard.setFiltersInStorage(new_filters);
```
```suggestion
```

trasher commented:
Could you please rebase to fix conflicts?
